### PR TITLE
Amir Sedighi, Microservices Challenge 2

### PR DESCRIPTION
**Pull request by amirsedighi**
# B-Yond, The second challenge, Save the Humans

Bots started an invasion to our planet. Regarding the latest research they leave some patterns in their reviews on amazon. Here we provde a REST API on the top of a Machine Learning microservice. It supposed to finds this patterns...
We've downloaded and mixed some datasets from here: http://jmcauley.ucsd.edu/data/amazon/
That data is being used to prepare the dictionary.txt file. dictionary.txt is the experience we make for the machine about human reviews.

## Architecture

We've used Microservices architecture. The services is stateless. So you can easily bring more instances up to scale out.
This service supposed to provide a low-level API. So it needs to be deployed behind firewalls, HA and non-blocking high-level APIs.

## Getting Started

1. Compile and buid the jar file:
```
mvn clean install -DskipTests
```

You can run service by one of the following ways:
```
mvn spring-boot::run
```

OR

Starting the project by building a docker image and running it withing the container:
```
docker-compose up
```

Then access the API online document using the following address:
```
http://localhost:8080/swagger-ui.html
```

Note:
The learner needs to be trained. It takes a few minutes to be trained... Ones it trained, you can start calling the API.

### Prerequisites

The learner needs at least 2GB ram. It uses a dictionary that downloaded from here:
http://jmcauley.ucsd.edu/data/amazon/

And make sure in src/main/resources/application.properties, the following path, points the current address on your disk:
```
basePath=/Users/amirsedighi/Projects/unmanaged/byond/amaze-us/microservice/challenge-2/save-the-humans/
```

## Running the tests

To run the tests just run the following command:
```
mvn test
```

Note: 
The learner needs to be trained. It takes a couple of minutes. Ones it trained, the tests will be running one after one.

## Try It

Request:
```
curl -X POST "http://localhost:8080/smell" -H "accept: */*" -H "Content-Type: application/json" -d "{ \"review\": \"Hi, map saving function is just coming out. We received a lot of advice from customers and will consider them seriously. Our developing department will complete more practical features gradually. We appreciate all good ideas from every customer and hope we could work with customer together to make our product better\", \"suspectedWords\": [ \"amazing\", \"human\" ]}"
```

Response:
```
{
  "review": "hi  map saving function is just coming out  .   we received a lot of advice from customers and will consider them seriously  .   our developing department will complete more practical features gradually  .   we appreciate all good ideas from every customer and hope we could work with customer together to make our product better",
  "verbose": "Seems the reviewer is a human being.",
  "suspectedWords": [
    "amazing",
    "human"
  ],
  "matchingSimilarWords": [],
  "matchingWords": [],
  "isBot": false
}
```

So he/she is a human! because isBot is false. There is also a verbose field.

Note:
You can try it with other suspected words. You are not limited to "amazing" and "human" keywords. The learner also will try the similar words as well.

See another sample:
```
curl -X POST "http://localhost:8080/smell" -H "accept: */*" -H "Content-Type: application/json" -d "{ \"review\": \"Hi, map saving function is just coming out. It is exciting and we received a lot of advice from customers and will consider them seriously. Our developing department will complete more practical features gradually. We appreciate all good ideas from every awesome customer and hope we could work with customer together to make our product better\", \"suspectedWords\": [ \"amazing\", \"human\" ]}"
```

Response is:
```
{
  "review": "hi  map saving function is just coming out  .   it is exciting and we received a lot of advice from customers and will consider them seriously  .   our developing department will complete more practical features gradually  .   we appreciate all good ideas from every awesome customer and hope we could work with customer together to make our product better",
  "verbose": "There is a small likelihood. But it is OK. But still I think he/she is a human being.",
  "suspectedWords": [
    "amazing",
    "human"
  ],
  "matchingSimilarWords": [
    "awesome"
  ],
  "matchingWords": [],
  "isBot": false
}
```

While the service knows she/he is not a machine, but "awesome" also being marked by service! Read the verbose:
"verbose": "There is a small likelihood. But it is OK. But still I think he/she is a human being.",

## Deployment

The project is dockerized now. Just use docker for deployment.

## Health

Service information:
```
curl http://localhost:8080/actuator/info
```

Service health:
```
curl http://localhost:8080/actuator/health
```

The project is dockerized now. Just use docker for deployment.

## Built With

* [Spring Boot](https://spring.io/projects/spring-boot) - Spring Boot, the glue framework, provides DI, REST Controllers.
* [DL4J](http://deeplearning4j.org) - The deep learning framework
* [Maven](https://maven.apache.org/) - Dependency Management
* [Swagger](https://swagger.io/) - Used to document the API

## Authors

* **Amir Sedighi**

## License

This project is licensed under the MIT License - see the [LICENSE.md](LICENSE.md) file for details

### CODE REVIEW

File: microservice/challenge-2/HELP.md [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+docker-compose up
+```
+
+Then access the API online document using the following address:
+```
+http://localhost:8080/swagger-ui.html
+```
+
+Note:
+After running service it starts to train itself. It takes a fe
```
Comment (albertaouad) [inline microservice/challenge-2/HELP.md]: incomplete sentence ?

File: microservice/challenge-2/HELP.md [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+
+### Prerequisites
+
+The learner needs at least 2GB ram. It uses a dictionary that downloaded from here:
```
Comment (albertaouad) [inline microservice/challenge-2/HELP.md]: you missed that URL ?

**Comment by amirsedighi** [comment]
Hi Albert,

You right!  The message supposed to be like this:
*The learner needs to be trained. It takes a few minutes to be trained...
Ones it trained, you can start calling the API.*

File: microservice/challenge-2/HELP.md (hunk 1 above)
Comment (amirsedighi) [reply microservice/challenge-2/HELP.md]: You right Albert. The message supposed to be like the followings:
**The learner needs to be trained. It takes a few minutes to be trained... Ones it trained, you can start calling the API.**

File: microservice/challenge-2/HELP.md (hunk 2 above)
Comment (amirsedighi) [reply microservice/challenge-2/HELP.md]: The URL:
 http://jmcauley.ucsd.edu/data/amazon/

**Comment by amirsedighi** [comment]
@albertaouad  I've just added a couple of lines in Prerequisites as well:

And make sure in application.properties, the following path points the current address on your disk:

basePath=/Users/amirsedighi/Projects/unmanaged/byond/amaze-us/microservice/challenge-2/save-the-humans/

File: microservice/challenge-2/src/main/java/com/byond/savethehumans/dispatch/Dispatcher.java [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+
+    /*  Loads the default suspected words such as Amazing and Human */
+    @Value("#{'${defaultSuspectedWords}'.split(',')}")
+    private  Set<String> defaultSuspectedWords;
+
+    Set<String> allSuspectedWords = new HashSet<>();
+
+    public ContentResponse investigate(UserReview c) throws Exception {
+
+        if (!word2VecIsEnabled) {
```
Comment (soujiro32167) [inline microservice/challenge-2/src/main/java/com/byond/savethehumans/dispatch/Dispatcher.java]: Why would word2Vec not be enabled? The application can't run without it...

Comment (amirsedighi) [reply microservice/challenge-2/src/main/java/com/byond/savethehumans/dispatch/Dispatcher.java]: Good question. There was a Multiple-Classifier as well. Just beside W2V. I wanted to use it beside W2V to make more exciting results. In the contentResponse DTO, we had a set of tags that were coming from the classifier such as Automotive, Video, Music ... 
At that time I had:
```
word2VecIsEnabled=true
reviewClassifier=true
```
So I was able to make it like this:
```
word2VecIsEnabled=false
reviewClassifier=true
```
and train the classifier faster...
When I noticed the results aren't enough good, I just removed the classifier.
word2VecIsEnabled also can be removed.

**Comment by soujiro32167** [comment]
Hey Amir

>The learner needs at least 2GB ram

Does the amount of RAM depend on input? How would you scale to 1TB of reviews?

**Comment by amirsedighi** [comment]
> Hey Amir
> 
> > The learner needs at least 2GB ram
> 
> Does the amount of RAM depend on input? How would you scale to 1TB of reviews?

Hi Eli,

Great question! The main idea here is to process reviews on-the-fly. We wanted to make it real-time to detect the robots while they are posting the reviews.
We can also ingest the previews old reviews as well.

Regarding the performance we need,  we can start a few instances of the current service up. For example 10 instances. It is stateless. So all instances work the same.
We may use an ensemble learning approach (Having different learners with different partial datasets) that is also scalable. 

I assumed there is a high-level API on the top of the current service (cluster) which is responsible for messaging, queueing and security... That HL-API can be used for ingesting the 1TB previews reviews into the cluster.

Meantime, fortunately, DL4J is totally Spark friendly. It is not a big deal to train the learner (On the top of Apache Spark) with very very big datasets.

So we can have a cluster of the current service and feeding them with all reviews (1TB or more) endlessly. Hope I explained the idea well.

**Comment by soujiro32167** [comment]
Thanks Amir, that makes sense!

**Comment by bruno-morel** [comment]
Closing this, Thanks you Amir!

**Pull request closed without merging**